File: osiconnect/src/services/examenService.ts
import { api } from "@/services/api";

export type TipoExamen = "PARCIAL" | "FINAL" | "RECUPERACION" | "PRUEBA";

export interface ExamenDTO {
    id: string;
    titulo: string;
    descripcion: string | null;
    cursoId: string;
    fecha: string;
    tipo: TipoExamen;
    puntuacionMaxima: number;
    totalNotas?: number;
    promedioNotas?: number;
}

export interface CrearExamenDTO {
    titulo: string;
    descripcion?: string;
    fecha: string;
    tipo: TipoExamen;
    puntuacionMaxima?: number;
}

export interface NotaExamenDTO {
    id: string;
    examenId: string;
    examenTitulo: string;
    alumnoId: string;
    alumnoNombre: string;
    nota: number | null;
    puntuacionMaxima: number;
    observaciones: string | null;
    fechaCalificacion: string;
}

export interface RegistrarNotaDTO {
    alumnoId: string;
    nota: number | null;
    observaciones?: string;
}

// Obtener todos los exámenes del curso
export async function getExamenesCurso(cursoId: string): Promise<ExamenDTO[]> {
    const res = await api.get(`/curso/${cursoId}/examenes`);
    return res.data;
}

// Obtener próximos exámenes
export async function getProximosExamenes(cursoId: string): Promise<ExamenDTO[]> {
    const res = await api.get(`/curso/${cursoId}/examenes/proximos`);
    return res.data;
}

// Obtener exámenes pasados
export async function getExamenesPasados(cursoId: string): Promise<ExamenDTO[]> {
    const res = await api.get(`/curso/${cursoId}/examenes/pasados`);
    return res.data;
}

// Crear examen (profesor/admin)
export async function crearExamen(cursoId: string, examen: CrearExamenDTO): Promise<ExamenDTO> {
    const res = await api.post(`/curso/${cursoId}/examenes`, examen);
    return res.data;
}

// Actualizar examen (profesor/admin)
export async function actualizarExamen(cursoId: string, examenId: string, examen: CrearExamenDTO): Promise<ExamenDTO> {
    const res = await api.put(`/curso/${cursoId}/examenes/${examenId}`, examen);
    return res.data;
}

// Eliminar examen (profesor/admin)
export async function eliminarExamen(cursoId: string, examenId: string): Promise<void> {
    await api.delete(`/curso/${cursoId}/examenes/${examenId}`);
}

// Obtener notas de un examen (profesor/admin)
export async function getNotasExamen(cursoId: string, examenId: string): Promise<NotaExamenDTO[]> {
    const res = await api.get(`/curso/${cursoId}/examenes/${examenId}/notas`);
    return res.data;
}

// Obtener mis notas (alumno)
export async function getMisNotasExamenes(cursoId: string): Promise<NotaExamenDTO[]> {
    const res = await api.get(`/curso/${cursoId}/examenes/mis-notas`);
    return res.data;
}

// Obtener mi promedio (alumno)
export async function getMiPromedioExamenes(cursoId: string): Promise<number> {
    const res = await api.get(`/curso/${cursoId}/examenes/mi-promedio`);
    return res.data;
}

// Registrar nota individual (profesor/admin)
export async function registrarNota(cursoId: string, examenId: string, nota: RegistrarNotaDTO): Promise<NotaExamenDTO> {
    const res = await api.post(`/curso/${cursoId}/examenes/${examenId}/notas`, nota);
    return res.data;
}

// Registrar múltiples notas (profesor/admin)
export async function registrarNotasMultiples(cursoId: string, examenId: string, notas: RegistrarNotaDTO[]): Promise<void> {
    await api.post(`/curso/${cursoId}/examenes/${examenId}/notas/multiple`, notas);
}

// Eliminar nota (profesor/admin)
export async function eliminarNota(cursoId: string, notaId: string): Promise<void> {
    await api.delete(`/curso/${cursoId}/examenes/notas/${notaId}`);
}

File: backend/src/main/java/com/infocurso/backend/service/ExamenService.java
package com.infocurso.backend.service;

import com.infocurso.backend.dto.*;
import com.infocurso.backend.entity.*;
import com.infocurso.backend.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ExamenService {

    private final ExamenRepository examenRepository;
    private final NotaExamenRepository notaExamenRepository;
    private final CursoRepository cursoRepository;
    private final UsuarioRepository usuarioRepository;

    // Obtener todos los exámenes de un curso
    public List<ExamenDTO> getExamenesCurso(UUID cursoId) {
        return examenRepository.findByCursoIdOrderByFechaAsc(cursoId)
                .stream()
                .map(e -> {
                    int totalNotas = e.getNotas().size();
                    Double promedio = notaExamenRepository.calcularPromedioExamen(e.getId());
                    return ExamenDTO.from(e, totalNotas, promedio);
                })
                .toList();
    }

    // Obtener próximos exámenes
    public List<ExamenDTO> getProximosExamenes(UUID cursoId) {
        return examenRepository.findProximosExamenes(cursoId)
                .stream()
                .map(ExamenDTO::from)
                .toList();
    }

    // Obtener exámenes pasados
    public List<ExamenDTO> getExamenesPasados(UUID cursoId) {
        return examenRepository.findExamenesPasados(cursoId)
                .stream()
                .map(e -> {
                    int totalNotas = e.getNotas().size();
                    Double promedio = notaExamenRepository.calcularPromedioExamen(e.getId());
                    return ExamenDTO.from(e, totalNotas, promedio);
                })
                .toList();
    }

    // Crear examen
    public ExamenDTO crearExamen(UUID cursoId, CrearExamenDTO dto) {
        Curso curso = cursoRepository.findById(cursoId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Curso no encontrado"));

        Examen examen = Examen.builder()
                .titulo(dto.titulo())
                .descripcion(dto.descripcion())
                .curso(curso)
                .fecha(dto.fecha())
                .tipo(dto.tipo())
                .puntuacionMaxima(dto.puntuacionMaxima() != null ? dto.puntuacionMaxima() : 10.0)
                .build();

        examen = examenRepository.save(examen);
        return ExamenDTO.from(examen);
    }

    // Actualizar examen
    public ExamenDTO actualizarExamen(UUID examenId, CrearExamenDTO dto) {
        Examen examen = examenRepository.findById(examenId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Examen no encontrado"));

        examen.setTitulo(dto.titulo());
        examen.setDescripcion(dto.descripcion());
        examen.setFecha(dto.fecha());
        examen.setTipo(dto.tipo());
        if (dto.puntuacionMaxima() != null) {
            examen.setPuntuacionMaxima(dto.puntuacionMaxima());
        }

        examen = examenRepository.save(examen);
        return ExamenDTO.from(examen);
    }

    // Eliminar examen
    public void eliminarExamen(UUID examenId) {
        if (!examenRepository.existsById(examenId)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Examen no encontrado");
        }
        examenRepository.deleteById(examenId);
    }

    // Obtener notas de un examen
    public List<NotaExamenDTO> getNotasExamen(UUID examenId) {
        return notaExamenRepository.findByExamenId(examenId)
                .stream()
                .map(NotaExamenDTO::from)
                .toList();
    }

    // Obtener mis notas de exámenes (para alumno)
    public List<NotaExamenDTO> getMisNotasExamenes(UUID alumnoId, UUID cursoId) {
        return notaExamenRepository.findByAlumnoIdAndExamenCursoId(alumnoId, cursoId)
                .stream()
                .map(NotaExamenDTO::from)
                .toList();
    }

    // Obtener promedio de un alumno
    public Double getPromedioAlumno(UUID alumnoId, UUID cursoId) {
        return notaExamenRepository.calcularPromedioAlumno(alumnoId, cursoId);
    }

    // Registrar nota individual
    public NotaExamenDTO registrarNota(UUID examenId, RegistrarNotaDTO dto) {
        Examen examen = examenRepository.findById(examenId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Examen no encontrado"));

        Usuario alumno = usuarioRepository.findById(dto.alumnoId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Alumno no encontrado"));

        // Buscar si ya existe una nota
        NotaExamen nota = notaExamenRepository.findByExamenIdAndAlumnoId(examenId, dto.alumnoId())
                .orElse(null);

        if (nota == null) {
            nota = NotaExamen.builder()
                    .examen(examen)
                    .alumno(alumno)
                    .nota(dto.nota())
                    .observaciones(dto.observaciones())
                    .build();
        } else {
            nota.setNota(dto.nota());
            nota.setObservaciones(dto.observaciones());
        }

        nota = notaExamenRepository.save(nota);
        return NotaExamenDTO.from(nota);
    }

    // Registrar múltiples notas
    public void registrarNotasMultiples(UUID examenId, List<RegistrarNotaDTO> notas) {
        for (RegistrarNotaDTO dto : notas) {
            registrarNota(examenId, dto);
        }
    }

    // Eliminar nota
    public void eliminarNota(UUID notaId) {
        if (!notaExamenRepository.existsById(notaId)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Nota no encontrada");
        }
        notaExamenRepository.deleteById(notaId);
    }
}
